HDDS-8128. Deduplicate the ops in RDBBatchOperation.

szetszwo opened the pull request:
## What changes were proposed in this pull request?

OM rocksdb uses a lot of space.

This PR adds a cache in RDBBatchOperation for caching the put ops.  When there are multiple ops with the same key, it will overwrite the previous op in the cache so the previous op won't be written into RocksDB.  In a test, the cache can reduce the disk usage in OM rocksdb of a single upload with 1000 parts from 755MB to 23MB.

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-8128

## How was this patch tested?

The is an optimization.  The existing test already has covered.  Also have tested this with https://github.com/apache/ozone/compare/ozone-1.3...szetszwo:ozone:multipartUpload manually.

Galsza commented on hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/utils/db/RDBBatchOperation.java:
```diff
@@ -21,15 +21,193 @@
 public class RDBBatchOperation implements BatchOperation {
+  static final Logger LOG = LoggerFactory.getLogger(RDBBatchOperation.class);
 
+  static void debug(Supplier<String> message) {
```
I think this method should be private. Are we going to use it anywhere else?

szetszwo replied:
Sure.

Galsza commented on hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/utils/db/RDBBatchOperation.java:
```diff
@@ -21,15 +21,193 @@
+  class PutOpCache {
+    class FamilyCache {
+      private final ColumnFamily family;
+      /** A (key -> value) map. */
```
```suggestion
      /** 
       * The map used to store the key and value byte array pairs.
       * <p>
       * {@link ByteArray} is used for the keys to avoid incorrect map 
       * operations.
       */
```

szetszwo replied:
Thanks for the suggestion.  Will update it.

Galsza commented on hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/utils/db/RDBBatchOperation.java:
```diff
@@ -21,15 +21,193 @@
+      void batchPut() throws IOException {
+        for (Map.Entry<ByteArray, byte[]> op : putOps.entrySet()) {
+          family.batchPut(writeBatch, op.getKey().bytes, op.getValue());
+        }
```
If I understand correctly, the batchSize, putCount etc. values should be reset once the batchPut operation is actually invoked. Should it be done here?

szetszwo replied:
A batch is only committed once.   Will add a precondition to make sure it is indeed the case.

Galsza replied:
Thank you. I misunderstood the code at first and didn't see that a new family cache is created once the changes are commited. It's looking good to me now.

szetszwo replied:
No problem.  The batches are single use objects.

Galsza commented on hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/utils/db/RDBBatchOperation.java:
```diff
@@ -21,15 +21,193 @@
+      }
+
+      void put(byte[] key, byte[] value) {
+        batchSize += key.length + value.length;
```
If batchSize is not reset anywhere this is going to overflow. Also, could this overflow even with a reset included? (Will cached put operations ever reach 2 gigs in size?)

szetszwo replied:
Good point.  We should use `long` for sizes.

Galsza commented on hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/utils/db/RDBBatchOperation.java:
```diff
@@ -21,15 +21,193 @@
+    }
+
+    /** A (family name -> {@link FamilyCache}) map. */
+    private final Map<String, FamilyCache> map = new HashMap<>();
```
I suggest renaming the variable to nameToCacheMap for easier understanding.

szetszwo replied:
Sure.

Galsza reviewed (changes requested):
@szetszwo Thank you for working on this patch. I've left a few comments.

Please fix the batchSize resetting, the rest are nitpicks.

Galsza reviewed (approved):
Thank you for the changes, looks good to me.

adoroszlai commented:
> force-pushed was to sync with master

Thanks for the info.  Sync with `master` can also be done by merging instead of rebasing.  The commits in the PR will be squashed anyway, so linearity doesn't matter.

szetszwo commented:
@adoroszlai , thanks for the advices.  I tend to prefer rebase over merge since it is easier to see the new code.  I agree that when everything works fine, rebase or merge do not matter.  When something does not work, rebase seems easier to trace the problem.

adoroszlai commented:
> I tend to prefer rebase over merge since it is easier to see the new code.

How does the reviewer see what changes were made to the PR since the previous round of review, i.e. to address the review comments?

szetszwo commented:
If we rebase but not squash, the comments in Github may still work.

Also, I usually keep the old branch https://github.com/szetszwo/ozone/tree/HDDS-8128b .  Reviewers may download the patches and the run diff locally (this is like posting patches on the JIRA).

szetszwo commented:
Indeed, a previous comment still show in https://github.com/apache/ozone/pull/4424/files .

The other comments become outdated.  I guess it is because the code has changed.  It has nothing to do with merge or rebase.

adoroszlai commented:
When posting patches to Jira, old ones are usually kept.  Force-pushing is like keeping only the latest patch.  Keeping around old branches helps, but you have to let others know about them, and even then it's still not as streamlined as  incremental commits.

You are right that comments are still visible after force-push.  However, it is hard to see the state of code they were referring to and how code was changed in response to them.

https://www.freecodecamp.org/news/optimize-pull-requests-for-reviewer-happiness#request-a-review

szetszwo commented:
For merging, the downloaded patch file could be very messy since it may include the merged code from other commit.  This is another reason I prefer rebase over merge.

> ... However, it is hard to see the state of code they were referring to and how code was changed in response to them.

Since the old branch is still around, we still can see the newer commits addressing the comments.

duongkame commented on hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/utils/db/RDBBatchOperation.java:
```diff
@@ -18,18 +18,225 @@
+
+    @Override
+    public int hashCode() {
+      return Arrays.hashCode(bytes);
```
This hashcode can/should be calculated at the constructor and cached.

szetszwo replied:
Sure.

duongkame commented on hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/utils/db/RDBBatchOperation.java:
```diff
@@ -18,18 +18,225 @@
 /**
  * Batch operation implementation for rocks db.
  */
 public class RDBBatchOperation implements BatchOperation {
```
It can be a good idea to put a comment stating this is not thread-safe and designed for single-thread use.

szetszwo replied:
Note that this is an existing class (not added by this change).  However, we could add the javadoc here.

duongkame replied:
Thanks for updating it. Before this class is just a proxy to `org.rocksdb.WriteBatch`, but now it has some internal logic that's not intended for multi-thread.

szetszwo replied:
Hey, `org.rocksdb.WriteBatch` itself is not thread safe, see https://github.com/facebook/rocksdb/wiki/Basic-Operations
> However other objects (like Iterator and WriteBatch) may require external synchronization. If two threads share such an object, they must protect access to it using their own locking protocol.

Indeed, making a *batch* operation thread safe does not sound right since a batch is supposed to be sequential.  Making it thread safe decreases the performance.

duongkame commented on hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/utils/db/RDBBatchOperation.java:
```diff
@@ -39,26 +246,38 @@ public RDBBatchOperation(ManagedWriteBatch writeBatch) {
       throws IOException {
+    debug(() -> String.format("%s: commit-with-writeOptions %s",
+        name, opCache.getCommitString()));
+    opCache.prepareBatchWrite();
```
I'd just pass `writeBatch` to `opCache.prepareBatchWrite()` to remove the nested dependency. `OpCache` can be considered being outside with a better name, e.g. `BatchOperationCache`.

szetszwo replied:
I guess you mean changing the classes to `static`?  I actually start with `static` and passing `writeBatch` around.  The `static` keywords were removed when printing the debug messages with the `name`.  It does not look good to pass also the `name` around.

duongkame reviewed (approved):
Thanks @szetszwo for the change. This looks like a nice improvement and maybe RocksDB should've done it internally. I only put a few minor comments inline.

szetszwo commented:
@duongkame , thanks a lot for reviewing this!  Just have pushed a new commit for addressing your comments.

swamirishi commented:
@szetszwo Did we see the disk usage go up because of the rocksdb wal? or did we see an sst redundant sst flush happen?

szetszwo commented:
WAL
